### DISCUSSION TITLE
Pass data with context object or with page queries ? Best practices ?

### DISCUSSION THREAD

**Discussion by AdrianGuery** [gatsbyjs/gatsby#34602] [Help]
Hello !
I'm currently working on a big Gatsby site (10,000+ pages) and we're trying to optimize build times. Because currently we take more than 8 minutes to build it locally (on high-performance machines).

The incremental build works well, it allows us to optimize time of the build of HTML pages. We managed to get off in less than 7 minutes.

But we need to get even higher build performance since the number of pages will grow exponentially. 

Thanks to an old discussion and an answer by @KyleAMathews I learned that the incremental build of Gatsby does not support filters on `allSomeType` page queries. And in addition, Gatsby's documentation does indicate that to get good performance at build, we have to "_Query only need fields in page queries_" ([src](https://www.gatsbyjs.com/docs/how-to/performance/improving-build-performance/)).

So we thought to use the context object to pass the data to the template, but in the webinar @graysonhicks indicates that we have to avoid absolutely the "context stuffing". And, in fact, the documentation also indicate that we have to "_Query only needed fields in createPages_" ([src](https://www.gatsbyjs.com/docs/how-to/performance/improving-build-performance/)).

But a template page is use more than 10 000 times, so I need to pass a lot of data to this template. So, how can I pass this data to the template without lose build performance and keep the incremental build ? What are the best practices?

**Comment by LekoArts** [comment]
Hi!

You'll want to have a the following setup:

In `gatsby-node` query only a unique identifier, ideally the `id` of the node itself and pass that to the template via context:

```js
const pageTemplate = require.resolve(`./src/templates/template.jsx`)

exports.createPages = async ({ graphql, actions }) => {
  const { createPage } = actions

  const result = await graphql(`
    {
      allData {
        nodes {
          id
          slug
        }
      }
    }
  `)

  result.data.allData.nodes.forEach(d => {
    createPage({
      path: `/${d.slug}`,
      component: pageTemplate,
      context: {
        id: d.id,
      }
    })
  })
}
```

In your template you then use that `id` for getting the exact node:

```graphql
query($id: String) {
  data(id: { eq: $id }) {
    name
  }
}
```

That's the fastest and best way you can query data.

**Reply by AdrianGuery** [reply]
Hi,

Thank you  for your quick answer.
I have another question.

In the case where we need a list of 5 recipes from the datalayer for example, do we have to :
 
1) Query the ids in the createPage like this `allRecipes(filter : {...}){ id }` and then get the recipes like bellow. (In this case the query will rerun each build no ?)
2) Query the ids into our source plugin and add it to the Recipe datalayer object and get them into our page query

**Reply by LekoArts** [reply]
In `gatsby-node` you'd do your `allRecipes` with the 5 recipes, the rest of the example I gave stays the same. Alternatively, yes, you could also already limit/filter at the sourcing level. But this would make the source plugin less flexible so if you later decide that you need to source more, it's fine to filter at the Gatsby level.

Unless you invalidate the cache/delete the cache between runs those queries will be cached and not rerun.

**Reply by AdrianGuery** [reply]
Thank you for your answer, so I pass the 5 recipes to the page in the context, like that ?
If I take the rest of the example : 

```

const pageTemplate = require.resolve(`./src/templates/template.jsx`)

exports.createPages = async ({ graphql, actions }) => {
  const { createPage } = actions

  const result = await graphql<{ allDuneYaml: { nodes: { id: string; name: string }[] } }>(`
    {
      allData {
        nodes {
          id
          slug
        }
      }
    }
  `)
  
    const fiveRecipesResult = await graphql<{ ... }>(`
    {
      allRecipes(filter : {...}) {
        nodes {
          id
          slug
          ...some data...
        }
      }
    }
  `)

  result.data.allData.nodes.forEach(d => {
    createPage({
      path: `/${d.slug}`,
      component: pageTemplate,
      context: {
        id: d.id,
        fiveRecipes: fiveRecipesResult
      }
    })
  })
}

```

**Reply by LekoArts** [reply]
I'm finding it hard to follow what you want to do exactly, if you can I'd appreciate a more complete example/repo.

You'll want to do the query for the five recipes on the `pageTemplate` query itself. In `createPages` if you can only query the list of IDs you need, pass that ID to the context and then use it in the template.

**Reply by ansmlc** [reply]
> Hi!
> 
> You'll want to have a the following setup:
> 
> In `gatsby-node` query only a unique identifier, ideally the `id` of the node itself and pass that to the template via context:
> 
> ```js
> const pageTemplate = require.resolve(`./src/templates/template.jsx`)
> 
> exports.createPages = async ({ graphql, actions }) => {
>   const { createPage } = actions
> 
>   const result = await graphql(`
>     {
>       allData {
>         nodes {
>           id
>           slug
>         }
>       }
>     }
>   `)
> 
>   result.data.allData.nodes.forEach(d => {
>     createPage({
>       path: `/${d.slug}`,
>       component: pageTemplate,
>       context: {
>         id: d.id,
>       }
>     })
>   })
> }
> ```
> 
> In your template you then use that `id` for getting the exact node:
> 
> ```graphql
> query($id: String) {
>   data(id: { eq: $id }) {
>     name
>   }
> }
> ```
> 
> That's the fastest and best way you can query data.

What if we have a lot of templates files and pages re-running the same queries again and again. I'm currently getting 40min+ build times due to a large number of slow query workers. Using createPages and for each node it says "Slow query took more than 15s". 

Would it be potentially helpful to query all (or part of) of data in gatsby-node and just forward it via pageContext?

**Reply by graysonhicks** [reply]
You will always be better/faster to run the queries primarily in the page template.  What does your page query look like?

**Reply by ansmlc** [reply]
> What does your page query look like?

This is the query in `templates/blogPostTemplate.js`:

```
export const query = graphql`
query {
  allMdx(
    sort: { fields: frontmatter___date, order: DESC }
  )     
  {
    nodes {
      excerpt
      timeToRead
      frontmatter {
        author
        category
        date(formatString: "MMMM DD, yyyy")
        featured
        metaDescription
        metaTitle
        slug
        tags
        title
        thumbnailAlt
        thumbnail {
          publicURL
          childImageSharp {
            gatsbyImageData
          }
        }
      } 
      body
    }
  }
  allCategoriesJson {
    nodes {
      name
      color
    }
  }
  allNewsCategoriesJson {
    nodes {
      name
      color
    }
  }
  allAuthorsJson {
    nodes {
      name
      thumbnail {
        childImageSharp {
          gatsbyImageData
        }
      }
    }
  }
  allNewsAuthorsJson {
    nodes {
      name
      thumbnail {
        childImageSharp {
          gatsbyImageData
        }
      }
    }
  }
}
`
```
```
const BlogPostTemplate = ({ data, pageContext }) => {
  let categoryData = '';
  let authorPath = '';
  if (pageContext.prefix === 'news') {
    categoryData = data.allNewsCategoriesJson
    authorPath = data.allNewsAuthorsJson
  } else {
    categoryData = data.allCategoriesJson
    authorPath = data.allAuthorsJson
  }

  const postData = data?.allMdx.nodes.filter(
    post => post.frontmatter.title === pageContext?.slug
  )?.[0]
  const nextPostData = data?.allMdx.nodes.filter(
    post => post.frontmatter.title === pageContext?.nextSlug
  )?.[0]
  const prevPostData = data?.allMdx.nodes.filter(
    post => post.frontmatter.title === pageContext?.previousSlug
  )?.[0]
  const authorData = authorPath?.nodes.filter(
    author => author.name === postData.frontmatter.author
  )?.[0]
```

**Reply by graysonhicks** [reply]
Ah, you are doing a lot of filtering via JS, which means the full query is run and the full result is stored.  You will want to use those `pageContext` values as filters in GraphQL instead and query individual `mdx` with the filter instead of `allMdx`.  That will make the query faster and the resulting data much smaller.

**Reply by ansmlc** [reply]
> Ah, you are doing a lot of filtering via JS, which means the full query is run and the full result is stored. You will want to use those `pageContext` values as filters in GraphQL instead and query individual `mdx` with the filter instead of `allMdx`. That will make the query faster and the resulting data much smaller.

Good point. The reason I'm getting all data is because child components use it to show related posts as well as show categories with (number of posts) labels, e.g. "Category 1 (22)" - so needed to count all the nodes.  I guess I'll have to give up on these features if I want a faster build.

But why "mdx" instead of "allMdx" - AFAIK the latter can be filtered via GraphQL just as well?

**Reply by graysonhicks** [reply]
You use `mdx` because in your filtering you are only returning one item.  Filter with GraphQL.  It will be faster build and smaller page-data.json files.  You can still do what you are trying to accomplish.

```javascript
export const query = graphql`
query BlogPost($slug: String, $nextSlug: String, $prevSlug: String) {
  thisPost: mdx(
    filter: { frontmatter: title: { eq: $slug }}
  ) {
    ..fields for post
  }
  nextPost: mdx(
    filter: { frontmatter: title: { eq: $nextSlug }}
  ) {
    ..fields for post
  }
  prevPost: mdx(
    filter: { frontmatter: title: { eq: $prevSlug }}
  ) {
    ..fields for post
  }
  allCategoriesJson {
    nodes {
      name
      color
    }
  }
  allNewsCategoriesJson {
    nodes {
      name
      color
    }
  }
  allAuthorsJson {
    nodes {
      name
      thumbnail {
        childImageSharp {
          gatsbyImageData
        }
      }
    }
  }
  allNewsAuthorsJson {
    nodes {
      name
      thumbnail {
        childImageSharp {
          gatsbyImageData
        }
      }
    }
  }
}
`
```

For the categories, you probably want to move those to two components `<Categories />` and `<NewsCategories />` that have static queries fetching only what they need.  Then use the pageContext in the BlogPostTemplate to determine which one to render.

```
// inside render return
...
pageContext.prefix === 'news' ? <NewsCategories /> : <Categories />
...
```
And inside those components each one would have a static query fetching only what it needs.  Maybe keep author at the template level though because you should also be filtering that with GraphQL as well.

**Reply by ansmlc** [reply]
> You use `mdx` because in your filtering you are only returning one item. Filter with GraphQL. It will be faster build and smaller page-data.json files. You can still do what you are trying to accomplish.
> ...

I love this. Huge help. Thanks!
